### ISSUE TITLE
IAsyncOperation array types in MIDL 

### ISSUE THREAD

**Issue by sichbo** [MicrosoftDocs/winrt-related#85]
Jeez, who's bugging you guys more than me lately, eh? 🙄 So here's something that ostensibly "should" work in MIDL 3.0 but raises a compilation error. A solution/reason why it doesn't work would be great to have in the intro here.

Busted MIDL:
```
runtimeclass Thing
{
	Windows.Foundation.IAsyncOperation&lt;Windows.Foundation.Collections.IVector&lt;String&gt;&gt; ListThings();
}
```

Emits:
```
error MIDL2025: [msg]syntax error [context]: expecting &gt; or , near "&gt;&gt;"
error MIDL2025: [msg]syntax error [context]: expecting &gt; or , near "("
```
Another scenario is IAsyncOperation&lt;Anything[]&gt;. It seems like MIDL 3.0 simply doesn't support nested generic types or arrays on any level. If you use IAsyncOperation with a basic type it's fine. 

I looked everywhere for an example MIDL returning an IAsyncOperation collection but came up empty. Any insights? or do we have to reinvent the wheel defining a custom collection wrapper every time we need to return an array asynchronously? Seems a bit much..

---
#### Document Details

⚠ *Do not edit this section. It is required for docs.microsoft.com ➟ GitHub issue linking.*

* ID: 2dacbdf2-172f-dd41-829f-94295cd0c621
* Version Independent ID: b511accc-e0b1-9b30-d044-231d5204024d
* Content: [Introduction to Microsoft Interface Definition Language 3.0 - Windows UWP applications](https://docs.microsoft.com/en-us/uwp/midl-3/intro)
* Content Source: [winrt-related-src/midl-3/intro.md](https://github.com/MicrosoftDocs/winrt-related/blob/docs/winrt-related-src/midl-3/intro.md)
* Product: **uwp**
* GitHub Login: @stevewhims
* Microsoft Alias: **stwhi**

**Comment by stevewhims** [comment]
The first issue turns out to be a whitespace issue. I can repro the issue but then if I add whitespace like this, it's fine:

```
Windows.Foundation.IAsyncOperation< Windows.Foundation.Collections.IVector< String > > ListThingsAsync();
```

This used to be an issue with C++ templates, too (not sure whether it still is). I got into the habit of adding a single space of padding inside all my angle brackets.

And, I've confirmed that it's not valid to use an array in that context (arrays are represented as a single parameter in metadata but as two tied parameters at the application binary interface layer). However, see whether [**IReferenceArray\<T\>**](/uwp/api/windows.foundation.ireferencearray_t_) can help you there.

I've added this info to the intro and to the troubleshooting topics.

Thanks!
~Steve
